File: Searching/binary search/python/binarySearch.py
def binarySearch(A,num,l,r):
    mid = l + (r-l)//2
    if r >= l:
        if A[mid] == num:
          return("Found")
        elif A[mid] > num:
            return binarySearch(A,num,l,mid-1)
        elif A[mid] < num:
            return binarySearch(A,num,mid+1,r)
    else: 
        return("Not Found")
   

#this non recursive function will return index as well. e is the element to be searched
def non_recursive_binary_search(a,n,e):
    i,j=0,n-1
    while i<=j:
        m=(i+j)//2
        if a[m]==e:
            return 'found at index ' + str(m)

        #seach in the right half
        if e>a[m]:
            i=m+1

        #search in the left half
        else:
            j=m-1

    return 'Element not found'


print("Enter the elements of the array in a single line: ")
a = list(map(int,input().split()))
num = int(input("Enter the number you want to search"))
print(binarySearch(a,num,0,len(a)))

#test the non recursive function and see if both the functions yield the same result
print(non_recursive_binary_search(a,len(a),num))

File: Searching/binary search/rust/binary_search.rs
fn binary_search(v : &Vec<i32>, value : i32) -> i32{
    let mut low = 0;
    let mut high = v.len()-1;

    while low <= high {

        let mid  = (low+high)/2;

        if value > v[mid]{
            low = mid+1;
        }
        else if value < v[mid] {
            high = mid-1;
        }else {
            return mid as i32;
        }
    }
    i32::MAX
}


fn main() {
    let mut v: Vec<i32> = vec![1, 2, 3, 4, 5, 9, 0];
    v.sort();
    let index = binary_search(&v, 9);
    if index != i32::MAX{
        println!("index of element is -> {}", index);
    }
    else {
        println!("Element not found");
    }
}
